### ISSUE TITLE
Typesafe Links

### ISSUE THREAD

**Issue by brillout** [brillout/vite-plugin-ssr#698]
### Description

I've a design in mind to make this work:

```tsx
// TypeScript complains that '/produt/42' isn't matching '/product/${string}' | '/' | '/some-other-route'
<Link href="/produt/42" />
```

React with 👍 if you want this.

**Comment by samuelstroschein** [comment]
Could a global type declaration for anchor tags `<a>` achieve typesafety instead of a custom link component?

**Comment by brillout** [comment]
I stand to be corrected, but I believe that isn't possible.

The idea is to have VPS code-gen `type Routes = '/product/${string}' | '/' | '/some-other-route'` while:
 - `Routes` is completely UI framework agnostic and can be used by the user/framework's `<Link>` implementation.
 -  The code-gen will be completely seamless (although Yarn users will have to add a line to their `tsconfig.json`). (But I've to check whether TypeScript doens't do something unexpected with the trick I've in mind.)

**Comment by samuelstroschein** [comment]
> Routes is completely UI framework agnostic and can be used by the user/framework's <Link> implementation.
 
That makes sense. Users are required to provide their own <Link> component.

**Comment by brillout** [comment]
From Discord:

> It would be very nice if it were possible to auto-generate types for routes out of the box, to be able to do something like `<Link href={routes.Product} parameters={[paramA, paramB]}`, with type-safety for all parameters and the route itself.

Yea I thought about this but, so far, I'm against this. Because it would require the user to give each and every route a name, which is pesky. And the ability to change the route of a page without having to update every `<Link>` is clearly useful, but not *that* useful (instead you can let TypeScript tell you what links you have to update which is an acceptable DX, I think).

**Comment by strlns** [comment]
> From Discord:
> 
> > It would be very nice if it were possible to auto-generate types for routes out of the box, to be able to do something like `<Link href={routes.Product} parameters={[paramA, paramB]}`, with type-safety for all parameters and the route itself.
> 
> Yea I thought about this but, so far, I'm against this. Because it would require the user to give each and every route a name, which is pesky. And the ability to change the route of a page without having to update every `<Link>` is clearly useful, but not _that_ useful (instead you can let TypeScript tell you what links you have to update which is an acceptable DX, I think).

Yes TS template strings are clearly less cluttered. This is better, didn't even think about that feature.
Now I'm curious how file-based routing works with i18n. Is this also a part of the idea? Haven't used i18n with VPS so far a at all.

**Comment by brillout** [comment]
Currenlty, i18n is being implemented programmatically, which provides a lot of flexibility, see https://vite-plugin-ssr.com/i18n.

And, yes, for typesafe links a declarative i18n interface will be needed. (Not only for providing typesafe i18n links, but also for a simpler interface that works for the vast majority of users.)

**Comment by strlns** [comment]
Makes senses, thank you for your reply. 

Programmatic i18n leaves most flexibility for routes (e.g. subdomain vs pathname prefix) and people can still share functions, components or e.g. React hooks to work with the API.

**Comment by quinnvaughn** [comment]
Would this also apply to navigate? 

remix-routes does this as $path so you'd just by default kill two birds with one stone, as you just call it with that function.

**Comment by brillout** [comment]
> Would this also apply to navigate?

Yes.

**Comment by micahjon** [comment]
Very much looking forward to using this!

It’d be awesome if in addition to type-safe pathnames, there was support (or at least a recommendation) for how to set and consume type-safe query parameters / hashes in the URL, similar to TanStack Router.

**Comment by brillout** [comment]
>  type-safe query parameters / hashes in the URL, similar to TanStack Router.

I agree.

**Comment by Blankeos** [comment]
Hi guys! I was doing some rough experiments to achieve some level of typesafety for `<a href={} />`, `navigate()` and consuming params (`useParams()`). I summarized it all here: https://github.com/blankeos/vike-routegen 👈.

It's a rough implementation that I'm separately running alongside `vike` and `vike-solid` for now, so I'm using `fs` to determine the routes, but pulling my discussion with @magne4000:

> You should be able to extract all Vike's routes from a Vite plugin, take a look at what I'm doing for vike-vercel: https://github.com/magne4000/vite-plugin-vercel/blob/bfe72c9c878b70eb3d8d282f0f809f19030fdfa3/packages/vike/src/plugins/routes.ts#L46

I figure that to do this, I would need to add the route generation script inside of the `vike` core itself. Then somehow integrate it with `vike-solid`, `vike-react`, and `vike-vue` extensions. Idk if this gets the ball rolling.

**Comment by brillout** [comment]
@Blankeos Thank you for paving the way. My thoughts on type safe routes are summed up at:
- https://github.com/vikejs/vike/issues/2253

I also have some neat new ideas I'll be writing down once I'm done with my current priorities.
